From 2ba23d33f0cbb479cbbf29c3bc318dccbde6aac1 Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Mon, 24 Jul 2023 17:52:09 +0200
Subject: [PATCH] arm64: dts: cn9131-cf-solidwan: switch cp0_phy0 to
 auto-negotiation

CP0 MDIO bus has 3 PHys connected, one on the CN9130 SoM, plus 2x on the
CN9131 SolidWAN Carrier. All PHYs are MV88E15 which only support 2
possible MDIO addresses: 0; 1.

The single PHY on SoM uses RGMII connection to the SoC, however the SoC
mac is reconfigured for using Serdes instead, which is routed to the
carrier.

All configuration commands to the 2x PHYs at address 0 receive their
response from the PHY on the SoM first, thereby masking any link-status
updates from other PHY - and preventing actual link-up for cp0 eth1.

As a workaround avoid MDIO and rely solely on SGMII auto-negotiation
between the MAC and the PHY for link-status and speed changes.

After power-on it is expected for U-Boot to send initial mdio
configuration commands to setup the PHY(s) as needed.

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 .../boot/dts/marvell/cn9130-cf-solidwan.dts   | 21 +++++++++++--------
 1 file changed, 12 insertions(+), 9 deletions(-)

diff --git a/arch/arm64/boot/dts/marvell/cn9130-cf-solidwan.dts b/arch/arm64/boot/dts/marvell/cn9130-cf-solidwan.dts
index efd1cf5ef906..4c81e76a0521 100644
--- a/arch/arm64/boot/dts/marvell/cn9130-cf-solidwan.dts
+++ b/arch/arm64/boot/dts/marvell/cn9130-cf-solidwan.dts
@@ -240,12 +240,12 @@ &cp0_mdio {
 	status = "okay";
 	pinctrl-names = "default";
 	pinctrl-0 = <&cp0_ge_mdio_pins>;
-	cp0_phy0: ethernet-phy@0 {
-		/* Green led blinks on activity, Orange LED on link */
-		//marvell,reg-init = <3 16 0 0x0064>;
-		marvell,reg-init = <3 16 0 0x1017>;
-		reg = <0>;
-	};
+
+	/*
+	 * SoM + Carrier both have PHY at address 0,
+	 * ignore MDIO and use SGMII Auto-Negotation between MAC and PHY.
+	 */
+
 	cp0_phy1: ethernet-phy@1 {
 		/* Green led blinks on activity, Orange LED on link */
 		//marvell,reg-init = <3 16 0 0x0064>;
@@ -284,11 +284,14 @@ &cp0_eth0 {
 
 /* SRDS #3 - SGMII 1GE on carrier board */
 &cp0_eth1 {
-	status = "okay";
 	phys = <&cp0_comphy3 1>;
-	phy = <&cp0_phy0>;
-	managed = "in-band-status";
+	/*
+	 * Due to mdio address conflict use SGMII Auto-Negotation
+	 * between MAC and PHY only, skipping MDIO.
+	 */
 	phy-mode = "sgmii";
+	managed = "in-band-status";
+	status = "okay";
 };
 
 /* SRDS #4 - USB 3.0 host on M.2 connector */
-- 
2.35.3

